orttraining/orttraining/python/orttraining_pybind_state.cc

```diff
@@ -357,16 +362,106 @@ void addObjectMethodsForTraining(py::module& m, ExecutionProviderRegistrationFn
       .def("reserve", [](std::vector<OrtValue>* v, const size_t len) { v->reserve(len); })
       .def("shrink_to_fit", [](std::vector<OrtValue>* v) { v->shrink_to_fit(); })
       .def("__len__", [](const std::vector<OrtValue>& v) { return v.size(); })
-      .def("__iter__", [](const std::vector<OrtValue>& v) {
-        return py::make_iterator(v.cbegin(), v.cend());
-      },
-           py::keep_alive<0, 1>())
+      .def(
+          "__iter__", [](const std::vector<OrtValue>& v) {
+            return py::make_iterator(v.cbegin(), v.cend());
+          },
+          py::keep_alive<0, 1>())
       .def("__getitem__", [](const std::vector<OrtValue>& v, const size_t idx) {
         return v.at(idx);
       })
       .def("dlpack_at", [](std::vector<OrtValue>* v, const size_t idx) {
         return py::reinterpret_steal<py::object>(ToDlpack(v->at(idx)));
-      });
+      })
+      .def("to_dlpack", [](const std::vector<OrtValue>& v, py::object to_tensor) -> py::list {
+
+        if (v.size() == 0)
+          return py::list();
+
+        py::list res;
+
+        PyObject* capsule;
+        PyObject* handle = to_tensor.ptr();
+        PyObject* obj;
+
+        py::gil_scoped_acquire acquire;
+
+        const Tensor& tensor = v[0].Get<Tensor>();
+
+        if (tensor.Location().device.Type() == OrtDevice::CPU) {
+          OrtDLManagedTensor ort_dlmanaged_tensor;
+          capsule = PyCapsule_New(&ort_dlmanaged_tensor.tensor, "dltensor", NULL);
```

Comment:
capsule  seems to be leaked after the whole function returns

Reply:
true